### DISCUSSION TITLE
Degree of automation for hierarchical edges

### DISCUSSION THREAD

**Discussion by Rojods** [kieler/KLighD#136] [Q&A] [answered]
Hello all! This is my first question around, so I wanna start by thanking you all for the amazing effort! :)

I have a question regarding the automatic drawings of edges. As disclaimer, I've not read the papers that discuss the ideas behind ELK and KlighD in detail, so I beg your pardon if the answer lies there (I accept references to the papers too :) ). My terminology is directly taken from [the ELK data structure section](https://www.eclipse.org/elk/documentation/tooldevelopers/graphdatastructure.html).

I am trying to visualize a KlighD model in which there are hierarchical edges, starting from a KGT file. Something like:

```
knode p1 {
  knode c1 {
    kport port1
    kedge (:port1 -> c2:port2)
  }
}

knode p2 {
  knode c2 {
    kport port2
  }
}
```

But the generated diagram does not contain the edge between `c1` and `c2` , crossing through `p1` and `p2`'s barriers, as visualized from KEITH 0.1.0. I am aware from #128 that KEITH is being deprecated as VSCode will take over, and maybe this has a better answer with the extension 0.2.1. But for the moment I did not find a reasonably easy up-to-date path to a working VSCode with graphics for the impatient. Or for the person with little time! The only extension in the VSCode marketplace as of 2022-07-18 is more of library than a extension, by self-claim.

I understand that from a holistic perspective it might be disadvantageous to support his type of model directly, as different users might have different expectations of what kind of diagram should pop out. But I wonder if there are some nice and still hidden options to be able to render this without manual elimination of long hierarchical edges? By manual I do imply a polynomial algorithm in time and space to do the ports and extra link creations in order to reduce long hierarchical edges to short hierarchical edges.

**Comment by NiklasRentzCAU** [comment, marked as answer]
Sorry for the late answer here, I forgot to activate the notifications for new discussions here and did not see this until now. 

The KLighD VS Code Extension is indeed only to be used as a library to be used by other extensions. But recently we published a first release of our KIELER VS Code extension, supporting the .kgt language among others. You can find that extension to view a diagram of your .kgt in VS Code in the [marketplace](https://marketplace.visualstudio.com/items?itemName=kieler.keith-vscode).

If we did not answer this yet when we met on FDL and for further reference, the answer to your question: To use the hierarchic edges, the ELK layout needs to be made conscious about the fact hierarchic layout should be used, in .kgt by providing this property at the top level graph:
```properties: org.eclipse.elk.hierarchyHandling=INCLUDE_CHILDREN```
See also [the ELK documentation for this option](https://www.eclipse.org/elk/reference/options/org-eclipse-elk-hierarchyHandling.html).

However, for this use case this will result in an incorrect rendering order for the hierarchic edge (will be drawn under the target node, I will create an issue for that). In general, hierarchic edges can get problematic in different ways, so avoiding that in general may give better results. Here your presented .kgt with the drawn edge, while avoiding a real hierarchic edge by routing the edge through invisible ports at the hierarchy levels, which you called the "manual" approach:
```
knode p1 {
    knode c1 {
        kport port1
        kedge (:port1 -> p1:p1p) {
            kpolyline // Do not draw an arrow head for this part of the edge
        }
    }
    kport p1p {
        size: height=0 width=0.1 // invisible hierarchy-breaking port
        krectangle {
            styles: invisible=true
        }
    }
    kedge (:p1p->p2:p2p) {
        kpolyline
    }
}

knode p2 {
    knode c2 {
        kport port2
    }

    kport p2p {
        size: height=0 width=0.1
        krectangle {
            styles: invisible=true
        }
    }

    kedge (:p2p->c2:port2)
}

```

**Reply by Rojods** [reply]
I had no idea you could do invisible ports. This is a life saving fact! Thanks a lot for letting me know!

**Reply by Rojods** [reply]
For the hierarchy, I agree I'd rather avoid them; but my usual problems inevitably fall into a hierarchical structure with less than trivial arcs connecting stuff. But at least with this new-found knowledge I can make it less cluttered.
